### DISCUSSION TITLE
DotNetCoreBuild: DLL being used by another process, build fails

### DISCUSSION THREAD

**Discussion by fbd** [cake-build/cake#3041] [Q&A] [answered]
Using dotnet 5.0.102 and cake.tool 0.38.5.  Building via cake results in:

`CSC : error CS2012: Cannot open 'C:\src\myapplication\obj\Release\net5.0\Application.dll' for writing -- 'The process cannot access the file 'C:\src\myapplication\obj\Release\net5.0\Application.dll' because it is being used by another process.' [C:\src\myapplication\application.csproj]`

about every other time I run `dotnet cake`.  However if I do something like `dotnet build; dotnet build; dotnet build; dotnet build; dotnet build; dotnet build;` (i.e. build directly multiple times in a row) I get no error.  That is, it seems like cake is doing something else I am not aware of or doing something slightly different, however the build script begins like this:

```
Task("Clean")
    .Does<BuildParameters>((parameters) =>
{
    CleanDirectory(workerbin);
    CleanDirectory(apibin);
    CleanDirectory(nugetDir);
});

Task("Build")
    .IsDependentOn("Clean")
    .Does<BuildParameters>((parameters) =>
{
     var settings = new DotNetCoreBuildSettings
    {
        Framework = "net5.0",
        Configuration = configuration,
        NoRestore = false,
        NoIncremental = true,
    };

    DotNetCoreBuild(solution, settings);
});
```

The directories being cleaned are in the root of the working directory, while the code being built and the solution file are at "./src/" so the CleanDirectory should not be touching anything in the src folder.  I have also tried changing the NoRestore, NoIncremental settings and removing them.  This happens both locally and on my build machine, and also when I run the command from as administrator.  I also tried to explicitly set the WorkingDirectory, same issue.

Has anyone else been experiencing something like this?

**Comment by patriksvensson** [comment, marked as answer]
@fbd Cake does not do anything special. This is related to the dotnet process not shutting down properly.
Microsoft added the `dotnet build-server` command just for this.

Try adding this to your script:

```csharp
Teardown(ctx => {
    Information("Shutting down .NET core SDK tooling...");
    DotNetCoreBuildServerShutdown(
        new DotNetCoreBuildServerSettings {
            MSBuild = true
        });
});
```

**Comment by fbd** [comment]
Thanks for the help!

Although your suggested fixed increased the odds of success (not scientifically measured) I do encounter this error occasionally, but a direct call to `dotnet build` never seems to be an issue (even though I call it couple times in a row).  It also looks like the failure happens when the last of the projects in the solution are being built.  So for example project A is built first and then B (which references A), then B fails because the dotnet process that was building A has not completed (I am guessing).  I guess there is one difference, `dotnet build` is invoked directly from the terminal and the other indirectly (i.e. Cake invokes `dotnet build`).  I suppose this is not necessarily related to Cake, but I have tried to emulate this by creating a Powershell script that calls `Invoke-Expression` to run `dotnet build` multiple times, but that works just fine.

**Reply by augustoproiete** [reply]
Is it possible you have any post-build event doing work on any project (or any MSBuild targets hooking up to the end of the build)?

Cake is just calling `dotnet build`, thus there should be no difference:

```powershell
Resolved tool to path C:/Program Files/dotnet/dotnet.exe
Executing: "C:/Program Files/dotnet/dotnet.exe" build "NameOfYourSolution.sln" --framework net5.0 --configuration Release --no-incremental
```

You might want to try to run the exact command-line that Cake runs on your PowerShell tests. You can see what Cake runs by setting the verbosity to `diagnostic` e.g. `dotnet cake build.cake --verbosity=diagnostic`

**Comment by fbd** [comment]
Thanks for your help!

This is definitely not related to Cake like you say, I get those errors if I explicitly use the `--framework` flag.  It is strictly speaking not needed since the TargetFramework is defined in all my .csproj files.  Removing the `Framework = "net5.0"` from the build script seems to be enough and works for me.  Hopefully this discussion will be of use to somebody else with the same problem.

Thanks again.

**Reply by pascalberger** [reply]
Glad you got it working and thanks for letting is know
